### PR TITLE
[diag] add network diagnostic data definition and parsing

### PR DESCRIPTION
**Pull request by ZhangLe2016**
supported TLVs:
- Ipv6addressList
- ChildIpv6AddressList
- Mode
- ChildTable
- LeaderData
- Route64
- MacCounters
- Eui64
- ExtMacAddress
- MacAddress

### CODE REVIEW

File: include/commissioner/network_diag_data.hpp [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+    ByteArray                    mExtMacAddress;
+    uint16_t                     mMacAddress = 0;
+    ModeData                     mMode;
+    Route64                      mRoute64;
+    LeaderData                   mLeaderData;
+    Ipv6AddressList              mIpv6AddressList;
+    std::vector<ChildTableEntry> mChildTable;
+    ByteArray                    mEui64;
+    ByteArray                    mTlvTypeList;
+    ChildIpv6AddressList         mChildIpv6AddressList;
```
Comment (wgtdkp) [inline include/commissioner/network_diag_data.hpp:162]:
```suggestion
    int mChildRloc16;
    std::vector<std::string>         mChildAddrs;
```

And remove the `ChildIpv6AddressList` class

File: include/commissioner/network_diag_data.hpp [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+ *    POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file defines the types of Thread Network Diagnostic TLVs used for network diagnostics.
+ */
+
+#ifndef OT_COMM_NETWORK_DIAG_TLVS_HPP_
+#define OT_COMM_NETWORK_DIAG_TLVS_HPP_
```
Comment (wgtdkp) [inline include/commissioner/network_diag_data.hpp:35]:
```suggestion
#ifndef OT_COMM_NETWORK_DIAG_DATA_HPP_
#define OT_COMM_NETWORK_DIAG_DATA_HPP_
```

File: include/commissioner/network_diag_data.hpp [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+#include <cstddef>
+#include <cstdint>
+#include <stdbool.h>
+#include <string>
+#include <vector>
+
+#include "defines.hpp"
+#include "error.hpp"
+
+#define CHILD_TABLE_ENTRY_BYTES 3
```
Comment (wgtdkp) [inline include/commissioner/network_diag_data.hpp:46]: We prefer `static constexpr int` over `#define`
```suggestion
static constexpr int CHILD_TABLE_ENTRY_BYTES = 3;
```

Please update for all constants and move them into the ot::commissioner:: namespace

File: include/commissioner/network_diag_data.hpp [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+ * `mPresentFlags`.
+ */
+struct NetDiagData
+{
+    ByteArray                    mExtMacAddress;
+    uint16_t                     mMacAddress = 0;
+    ModeData                     mMode;
+    Route64                      mRoute64;
+    LeaderData                   mLeaderData;
+    Ipv6AddressList              mIpv6AddressList;
```
Comment (wgtdkp) [inline include/commissioner/network_diag_data.hpp:158]:
```suggestion
    std::vector<std::string>              mAddrs;
```

File: include/commissioner/network_diag_data.hpp (hunk 1 above)
Comment (ZhangLe2016) [reply include/commissioner/network_diag_data.hpp:162]: no, we cannot separate the `ChildIpv6AddressList` to `mChildRloc16` and `mChildAddrs` here. Because a Thread device may have more than one child, your change cannot support it.

Comment (wgtdkp) [reply include/commissioner/network_diag_data.hpp:162]: I see why you want the `ChildIpv6AddressList` now, you want to use a single present flag for a single member field, right? I think it's okay for this purpose. But still you don't need the `Ipv6AddressList` class, pleae delete that one.

Besides, I think the name "ChildAddressInfo" is better than ChildIpv6AddressList as it includes the RLOC integer.

File: include/commissioner/network_diag_data.hpp [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+    std::vector<RouteDataEntry> mRouteData;
+};
+
+/**
+ * @brief Child IPv6 Address List
+ */
+struct ChildIpv6AddressList
+{
+    uint16_t        mRloc16 = 0;
+    Ipv6AddressList mIpv6AddressList;
```
Comment (wgtdkp) [inline include/commissioner/network_diag_data.hpp:125]: How can we identify which Child has those address? Should there be a `Child ID` or something similar?

File: include/commissioner/network_diag_data.hpp (hunk 4 above)
Comment (ZhangLe2016) [reply include/commissioner/network_diag_data.hpp]: done

File: include/commissioner/network_diag_data.hpp (hunk 3 above)
Comment (ZhangLe2016) [reply include/commissioner/network_diag_data.hpp]: done

File: include/commissioner/network_diag_data.hpp (hunk 5 above)
Comment (ZhangLe2016) [reply include/commissioner/network_diag_data.hpp]: we can identify the child with rloc16

File: include/commissioner/network_diag_data.hpp (hunk 1 above)
Comment (ZhangLe2016) [reply include/commissioner/network_diag_data.hpp]: I think the previous definition of the `ChildIpv6AddressList ` is incorrect.
it should be defined as follow:
```
std::vector<ChildIpv6AddressInfo> mChildIpv6AddrsInfoList;
```
`ChildIpv6AddressInfo` represents the address information recorded by a child of the device, including its RLOC (Routing Locator) and addresses.  std::vector<ChildIpv6AddressInfo> records the address information of all the children of the device.

File: include/commissioner/network_diag_data.hpp (hunk 5 above)
Comment (wgtdkp) [reply include/commissioner/network_diag_data.hpp]: Them how can we associate those addresses with a ChildTableEntry? There doesn't seem to be a rloc16 in ChildTableEntry...

Comment (ZhangLe2016) [reply include/commissioner/network_diag_data.hpp]: Although we can use `rloc16 & 0x1FF` to generate the `child ID` which can match with `ChildTableEntry `'s address, in practical applications, we still need to use RLOC16 to identify nodes in the network. In other words, the child ID in the Child Table Entry and the router ID of the device itself are used to generate the RLOC16 of that child.

File: src/library/commissioner_impl.cpp [hunk 6]
```diff
@@ -1955,6 +1955,288 @@ ByteArray CommissionerImpl::GetCommissionerDatasetTlvs(uint16_t aDatasetFlags)
...
+
+    while (offset < length)
+    {
+        ChildTableEntry entry;
+        VerifyOrExit(offset + kChildTableEntryBytes <= length,
+                     error = ERROR_BAD_FORMAT("premature end of Child Table"));
+        entry.mTimeout             = 1 << (((aBuf[offset] & 0xF8) >> 3) - 4);
+        entry.mIncomingLinkQuality = (aBuf[offset] & 0x06) >> 1;
+        entry.mChildId             = ((aBuf[offset] & 0x01) << 9) | aBuf[offset + 1];
+        SuccessOrExit(error = DecodeModeData(entry.mModeData, {aBuf.begin() + offset + 2, aBuf.begin() + offset + 3}));
```
Comment (wgtdkp) [inline src/library/commissioner_impl.cpp:2152]: Can you document the TLV value format here? It's hard to follow the code.

File: include/commissioner/network_diag_data.hpp [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+
+#include "defines.hpp"
+#include "error.hpp"
+
+namespace ot {
+
+namespace commissioner {
+
+static constexpr uint8_t kChildTableEntryBytes = 3;
+static constexpr uint8_t kIpv6AddressBytes     = 16;
```
Comment (wgtdkp) [inline include/commissioner/network_diag_data.hpp:51]: Is this necessary being part of the library API? Will an client app need to use this value? If no, please move this constant to the class where it is used.

File: include/commissioner/network_diag_data.hpp [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+#include <vector>
+
+#include "defines.hpp"
+#include "error.hpp"
+
+namespace ot {
+
+namespace commissioner {
+
+static constexpr uint8_t kChildTableEntryBytes = 3;
```
Comment (wgtdkp) [inline include/commissioner/network_diag_data.hpp:50]: Ditto, is this useful for client apps?

File: include/commissioner/network_diag_data.hpp [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+static constexpr uint8_t kIpv6AddressBytes     = 16;
+static constexpr uint8_t kLeaderDataBytes      = 8;
```
Comment (wgtdkp) [inline include/commissioner/network_diag_data.hpp:52]: Ditto, this doesn't seem useful for client apps.

**Review by wgtdkp — COMMENTED** [review]
Is it possible to add unit tests in commissioner_impl_test.cpp?

File: src/library/commissioner_impl.cpp (hunk 6 above)
Comment (ZhangLe2016) [reply src/library/commissioner_impl.cpp:2185]: done

File: src/library/commissioner_impl.cpp [hunk 10]
```diff
@@ -1955,6 +1955,288 @@ ByteArray CommissionerImpl::GetCommissionerDatasetTlvs(uint16_t aDatasetFlags)
...
+    Error                error;
+    size_t               length = aBuf.size();
+    ChildIpv6AddressInfo childIpv6AddrsInfo;
+
+    VerifyOrExit((length - kRloc16Bytes) % kIpv6AddressBytes == 0,
+                 error = ERROR_BAD_FORMAT("premature end of Child IPv6 Address"));
+    childIpv6AddrsInfo.mRloc16 = utils::Decode<uint16_t>(aBuf.data(), kRloc16Bytes);
+
+    SuccessOrExit(error = DecodeIpv6AddressList(childIpv6AddrsInfo.mAddrs, {aBuf.begin() + kRloc16Bytes, aBuf.end()}));
+    aChildIpv6AddressInfoList.emplace_back(childIpv6AddrsInfo);
```
[the remark this answers is not in the document]
Comment (ZhangLe2016) [reply src/library/commissioner_impl.cpp:2137]: no, the `DecodeChildIpv6AddressList ` may be call multiple times, and the reference value of `std::vector<ChildIpv6AddressInfo>` will be added new childIpv6AddrsInfo object at its end.

File: include/commissioner/network_diag_data.hpp [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+    uint16_t                          mMacAddress = 0;
+    ModeData                          mMode;
+    Route64                           mRoute64;
+    LeaderData                        mLeaderData;
+    std::vector<std::string>          mAddrs;
+    std::vector<ChildTableEntry>      mChildTable;
+    ByteArray                         mEui64;
+    ByteArray                         mTlvTypeList;
+    std::vector<ChildIpv6AddressInfo> mChildIpv6AddrsInfoList;
+    MacCounters                       mMacCounters;
```
[the remark this answers is not in the document]
Comment (ZhangLe2016) [reply include/commissioner/network_diag_data.hpp]: done.

File: src/library/commissioner_impl.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (wgtdkp) [reply src/library/commissioner_impl.cpp:2137]: Could you please point me to the multiple call sites of `DecodeChildIpv6AddressList`?

File: include/commissioner/network_diag_data.hpp [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+{
+    ByteArray                         mExtMacAddress;
+    uint16_t                          mMacAddress = 0;
+    ModeData                          mMode;
+    Route64                           mRoute64;
+    LeaderData                        mLeaderData;
+    std::vector<std::string>          mAddrs;
+    std::vector<ChildTableEntry>      mChildTable;
+    ByteArray                         mEui64;
+    ByteArray                         mTlvTypeList;
```
[the remark this answers is not in the document]
Comment (ZhangLe2016) [reply include/commissioner/network_diag_data.hpp]: the TLV named `TlvTypeList` is used for collecting the request/query TLVs' type, we just use it before executing request and query commands, so I think we can remove it from here.

File: include/commissioner/network_diag_data.hpp [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+    std::vector<ChildIpv6AddressInfo> mChildIpv6AddrsInfoList;
+    MacCounters                       mMacCounters;
+
+    /**
+     * Indicates which fields are included in the dataset.
```
[the remark this answers is not in the document]
Comment (ZhangLe2016) [reply include/commissioner/network_diag_data.hpp]: done.

File: src/library/commissioner_impl.cpp [hunk 14]
```diff
@@ -1955,6 +1955,288 @@ ByteArray CommissionerImpl::GetCommissionerDatasetTlvs(uint16_t aDatasetFlags)
...
+    ByteArray tlvTypes;
+
+    if (aDiagTlvFlags & NetDiagData::kExtMacAddressBit)
+    {
+        EncodeTlvType(tlvTypes, tlv::Type::kNetworkDiagExtMacAddress);
+    }
+    if (aDiagTlvFlags & NetDiagData::kMacAddressBit)
+    {
+        EncodeTlvType(tlvTypes, tlv::Type::kNetworkDiagMacAddress);
+    }
```
[the remark this answers is not in the document]
Comment (ZhangLe2016) [reply src/library/commissioner_impl.cpp]: done.  Thanks.

File: include/commissioner/network_diag_data.hpp (hunk 7 above)
Comment (ZhangLe2016) [reply include/commissioner/network_diag_data.hpp]: I think these const variables only be used by `commissioner_impl`, i have removed them to  `commissioner_impl`. Thanks.

File: include/commissioner/network_diag_data.hpp (hunk 9 above)
Comment (ZhangLe2016) [reply include/commissioner/network_diag_data.hpp]: done.

File: include/commissioner/network_diag_data.hpp [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+namespace ot {
+
+namespace commissioner {
+
+static constexpr uint8_t kChildTableEntryBytes = 3;
+static constexpr uint8_t kIpv6AddressBytes     = 16;
+static constexpr uint8_t kLeaderDataBytes      = 8;
+static constexpr uint8_t kMacCountersBytes     = 36;
+static constexpr uint8_t kRloc16Bytes          = 2;
+static constexpr uint8_t kRouterIdMaskBytes    = 8;
```
[the remark this answers is not in the document]
Comment (ZhangLe2016) [reply include/commissioner/network_diag_data.hpp]: done.

File: src/library/commissioner_impl.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (ZhangLe2016) [reply src/library/commissioner_impl.cpp:2137]: oh, I mistakenly thought that `tlvSet` in line 2029 could hold a map<type, List<Tlv>>, but it actually can only hold a map<type, Tlvptr>. This means I need to make some changes. Thanks for pointing that out!

File: include/commissioner/network_diag_data.hpp [hunk 16]
```diff
@@ -0,0 +1,170 @@
...
+struct NetDiagData
+{
+    ModeData                          mMode;
+    ByteArray                         mEui64;
+    ByteArray                         mExtMacAddress;
+    uint16_t                          mMacAddress = 0;
+    Route64                           mRoute64;
+    LeaderData                        mLeaderData;
+    MacCounters                       mMacCounters;
+    std::vector<std::string>          mAddrs;
```
Comment (wgtdkp) [inline include/commissioner/network_diag_data.hpp:145]: Some place we use "Addr" while some place we use "Address", it doesn't matter which one is used, but it would be better to make it consistent. Looking at existing code, "Addr" is always used for addresses. Can we update the code in PR to follow the existing naming? Thanks

File: src/library/commissioner_impl.cpp [hunk 17]
```diff
@@ -1955,6 +1962,314 @@ ByteArray CommissionerImpl::GetCommissionerDatasetTlvs(uint16_t aDatasetFlags)
...
+    return tlvTypes;
+}
+
+Error CommissionerImpl::DecodeNetDiagData(NetDiagData &aNetDiagData, const ByteArray &aPayload)
+{
+    Error       error;
+    tlv::TlvSet tlvSet;
+    NetDiagData diagData;
+    // Clear all data fields
+    diagData.mPresentFlags = 0;
```
Comment (wgtdkp) [inline src/library/commissioner_impl.cpp:2028]: nit: this can be removed given the default value for `mPresentFlags` is already `0`

File: include/commissioner/network_diag_data.hpp (hunk 5 above)
Comment (wgtdkp) [reply include/commissioner/network_diag_data.hpp]: This means the client app developers need to know that they need to get the child id from `rloc16 & 0x1FF`, this is a knowledge barrier. Can we include also the `ChildId` in the `ChildIpv6AddressInfo` class? Just as there is a `ChildId` in `ChildTableEntry`. Or provide a member method `getChildId() const {...}`

**Review by wgtdkp — COMMENTED** [review]
Thanks for resolving the comments!
In internal code review, we asked to mark comments resolved when you followed the suggestion in it and leave the other open. This is useful for reviewers to re-review the code and skip comments which has been resolved. It makes the review much easier.

Can you follow that practice in GitHub as well? For example, you can and should only close comments which you replied "done" :)

File: include/commissioner/network_diag_data.hpp (hunk 5 above)
Comment (ZhangLe2016) [reply include/commissioner/network_diag_data.hpp]: Added a new field named `ChildId` into the `ChildTableEntry`.

File: include/commissioner/network_diag_data.hpp (hunk 16 above)
Comment (ZhangLe2016) [reply include/commissioner/network_diag_data.hpp]: I see, I changed `addrs` to `address`

File: src/library/commissioner_impl.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (ZhangLe2016) [reply src/library/commissioner_impl.cpp:2144]: I added a new function named `GetTlvListByType` into Tlv.cpp module which can fetch the TLV list like `ChildIpv6AddressList` into a Tlv List, and I used this method to solve this problem.

File: include/commissioner/network_diag_data.hpp (hunk 16 above)
Comment (wgtdkp) [reply include/commissioner/network_diag_data.hpp]: No, if you look at include/commissioner/commissioner.hpp, "Addr" is used but not "Address"

**Comment by ZhangLe2016** [comment]
> Is it possible to add unit tests in commissioner_impl_test.cpp?

Looks like the unit test cases need to call the API functions defined into `commissioner.hpp`. For this PR, the new functions are defined in private, I am not sure if we can add corresponding test cases for them, do you have any suggestions?

**Comment by ZhangLe2016** [comment]
> > Is it possible to add unit tests in commissioner_impl_test.cpp?
> 
> Looks like the unit test cases need to call the API functions defined into `commissioner.hpp`. For this PR, the new functions are defined in private, I am not sure if we can add corresponding test cases for them, do you have any suggestions?

Added a new unit test case into `commissioner_impl_test.cpp`

File: src/library/tlv.hpp [hunk 18]
```diff
@@ -217,7 +217,9 @@ class Tlv
 
 Error  GetTlvSet(TlvSet &aTlvSet, const ByteArray &aBuf, Scope aScope = Scope::kMeshCoP);
 TlvPtr GetTlv(tlv::Type aTlvType, const ByteArray &aBuf, Scope aScope = Scope::kMeshCoP);
-bool   IsDatasetParameter(bool aIsActiveDataset, tlv::Type aTlvType);
+Error  GetTlvListByType(TlvList &aTlvList, const ByteArray &aBuf, tlv::Type aTlvType, Scope aScope = Scope::kMeshCoP);
+
+bool IsDatasetParameter(bool aIsActiveDataset, tlv::Type aTlvType);
```
Comment (wgtdkp) [inline src/library/tlv.hpp:222]: nit: is the newline intentional?
```suggestion
Error  GetTlvListByType(TlvList &aTlvList, const ByteArray &aBuf, tlv::Type aTlvType, Scope aScope = Scope::kMeshCoP);
bool   IsDatasetParameter(bool aIsActiveDataset, tlv::Type aTlvType);
```

File: src/library/commissioner_impl_test.cpp [hunk 19]
```diff
@@ -143,6 +145,43 @@ TEST(CommissionerImpl, NotImplementedApis)
...
+    EXPECT_EQ(diagData.mMode.mIsMtd, false);
+    EXPECT_EQ(diagData.mRoute64.mRouteData.size(), 9);
+    EXPECT_EQ(diagData.mAddrs.size(), 4);
+    EXPECT_EQ(diagData.mAddrs[0], "fd92:38a3:395d:1:f904:3dfe:b7b3:edf3");
+    EXPECT_EQ(diagData.mChildTable.size(), 3);
+    EXPECT_EQ(diagData.mChildTable[0].mChildId, 24);
+    EXPECT_EQ(diagData.mLeaderData.mRouterId, 33);
+    EXPECT_EQ(diagData.mChildIpv6AddrsInfoList.size(), 2);
+    EXPECT_EQ(diagData.mChildIpv6AddrsInfoList[0].mRloc16, 28696);
+    EXPECT_EQ(diagData.mChildIpv6AddrsInfoList[0].mChildId, 24);
```
Comment (wgtdkp) [inline src/library/commissioner_impl_test.cpp:182]: nit:
1. assert the child address as well
2. assert the second child informations as well

**Review by wgtdkp — APPROVED** [review]
Looks good to me overall, just minor comments!

File: src/library/commissioner_impl_test.cpp (hunk 19 above)
Comment (ZhangLe2016) [reply src/library/commissioner_impl_test.cpp:182]: done.

Comment (wgtdkp) [reply src/library/commissioner_impl_test.cpp:182]: Better assert rloc, address and childId for both two Children

Comment (ZhangLe2016) [reply src/library/commissioner_impl_test.cpp]: done.

**Comment by ZhangLe2016** [comment]
@jwhui Could you please review this PR? Thanks~~

**Review by jwhui — APPROVED** [review]
